tools/clang/tools/dxa/dxa.cpp

```diff
@@ -184,6 +189,76 @@ bool DxaContext::ExtractFile(const char *pName) {
   return printedAny;
 }
 
+bool DxaContext::AddPart(const char *pName, const char *pPartBinFilename) {
+  CComPtr<IDxcContainerReflection> pReflection;
+  CComPtr<IDxcBlobEncoding> pSource;
+  UINT32 partCount;
+
+  ReadFileIntoBlob(m_dxcSupport, StringRefUtf16(InputFilename), &pSource);
+  IFT(m_dxcSupport.CreateInstance(CLSID_DxcContainerReflection, &pReflection));
+  IFT(pReflection->Load(pSource));
+  IFT(pReflection->GetPartCount(&partCount));
+  IFTARG(strlen(pName) == 4);
+
+  const UINT32 matchName =
+      ((UINT32)pName[0] | ((UINT32)pName[1] << 8) | ((UINT32)pName[2] << 16) |
+       ((UINT32)pName[3] << 24));
+  for (UINT32 i = 0; i < partCount; ++i) {
+    UINT32 partKind;
+    IFT(pReflection->GetPartKind(i, &partKind));
+    if (partKind == matchName) {
+      printf("%s part is already exist\n", pName);
+      return false;
+    }
+  }
+  uint32_t fourcc = matchName;
+  CComPtr<IDxcBlobEncoding> pPartSource;
+  ReadFileIntoBlob(m_dxcSupport, StringRefUtf16(pPartBinFilename),
+                   &pPartSource);
+  CComPtr<IDxcContainerBuilder> pBuilder;
+  IFT(m_dxcSupport.CreateInstance(CLSID_DxcContainerBuilder, &pBuilder));
+  IFT(pBuilder->Load(pSource));
+  IFT(pBuilder->AddPart(fourcc, pPartSource));
+
+  CComPtr<IDxcOperationResult> pAssembleResult;
+  IFT(pBuilder->SerializeContainer(&pAssembleResult));
+
+  CComPtr<IDxcBlobEncoding> pErrors;
+  CComPtr<IDxcBlobUtf8> pErrorsUtf8;
+  pAssembleResult->GetErrorBuffer(&pErrors);
+  if (pErrors && pErrors->GetBufferSize() > 1) {
+    IFT(pErrors->QueryInterface(IID_PPV_ARGS(&pErrorsUtf8)));
+    printf("Errors or warnings:\n%s", pErrorsUtf8->GetStringPointer());
+  }
+
+  HRESULT status;
+  IFT(pAssembleResult->GetStatus(&status));
+  if (SUCCEEDED(status)) {
+    printf("Assembly succeeded.\n");
+    CComPtr<IDxcBlob> pContainer;
+    IFT(pAssembleResult->GetResult(&pContainer));
+    if (pContainer.p != nullptr) {
+      // Infer the output filename if needed.
+      if (OutputFilename.empty()) {
+        if (InputFilename == "-") {
+          OutputFilename = "-";
+        } else {
+          StringRef IFN = InputFilename;
+          OutputFilename = (IFN.endswith(".ll") ? IFN.drop_back(3) : IFN).str();
+          OutputFilename = (IFN.endswith(".bc") ? IFN.drop_back(3) : IFN).str();
+          OutputFilename += ".dxbc";
+        }
+      }
+
+      WriteBlobToFile(pContainer, StringRefUtf16(OutputFilename), DXC_CP_UTF8);
```

Comment:
The format is not UTF8, it's binary.  While this won't impact blobs that aren't considered text (have no encoding), this is misleading, since there's no way this should ever be text written as utf-8.  Should use `DXC_CP_ACP` here instead.  Note: this might not be the only place in dxa.cpp that makes this mistake.